### ISSUE TITLE
how can i get the covariance matrix after matching?

### ISSUE THREAD

**Issue by JokerJohn** [koide3/ndt_omp#20]
hello kode3, thanks  for your work! i have a question.when i use your ndt omp library for  localization, i do not know how to get the covariance matrix which can evaluate my result?

**Comment by idref** [comment]
Hello koide3, thanks for your great work!
I am working on the topic: covariance estimation of NDT-registrations. And I have the same question as JokerJohn.
How can I evaluate my registration results ? An estimated covariance matrix would be great for postprocessing. 

There is no implementation of AndreaCensis work (2007) (covariance estimation), right? 

In the ndt_omp_impl.hpp file (l: 935) is a function called calculateScore(). Is this the same implementation as the ndt score function of Magnusson2009 (negative logarithmic likelihood function) ?

**Comment by koide3** [comment]
> There is no implementation of AndreaCensis work (2007) (covariance estimation), right?

No, I didn't implement Censi's method and haven't seen any open source implementation of it. If you want to implement it you can find the Hessian at the following line. You would need a slight modification to access to it though.

https://github.com/koide3/ndt_omp/blob/399d61318a471758234e0d747ab4e407c74fcb76/include/pclomp/ndt_omp_impl.hpp#L113

> In the ndt_omp_impl.hpp file (l: 935) is a function called calculateScore(). Is this the same implementation as the ndt score function of Magnusson2009 (negative logarithmic likelihood function) ?

Yes, it is the negative log likelihood defined in Magnusson2009. But, in my experience, it didn't help to assess the registration result because it got often small for corrupted registration results.

**Comment by koide3** [comment]
Hi @idref ,

> Is this is the right place to take off the Hessian?

Yes, I think it's a correct place to get the final Hessian.

> All six eigenvalues are negative. Can you give me a hint why this is so or whether i can simply use the absolute vlaue of the eigenvalues with

If I remember correctly, this NDT implementation does maximization rather than minimization. If so, it's ok that the eigenvalues of the Hessian are negative. If you need only the magnitude of the largest eigenvalue, I think you can just take an absolute.
